lib/drivers/bayerContourNext.js

```diff
@@ -271,110 +272,134 @@ module.exports = function (config) {
+          var frame = hidDevice.nextPacket();
+          // TODO - we should ideally check the validity of the ASTM frame (sequence number, checksum etc) here as well for retries
+          //debug(debugFrame(frame.payload));
+          callback(err, frame.payload);
+        }
+      });
     });
   };
 
-  var receivePacket = function (timeout, parser, callback) {
+  var getASTMFrame = function (timeout, retries, callback) {
     var abortTimer = setTimeout(function () {
       clearInterval(listenTimer);
       debug('TIMEOUT');
-      callback('TIMEOUT', null);
+      callback('TIMEOUT');
     }, timeout);
 
     var listenTimer = setInterval(function () {
       hidDevice.receive(function(raw) {
-          var r = new Uint8Array(raw);
-          var startIndex = 0;
-          var rec = struct.unpack(r, 6, 'b', ['TYPE']);
-          var hasETB = _.indexOf(r, ASCII_CONTROL.ETB, startIndex);
-          if (hasETB !== -1) {
-            clearTimeout(abortTimer);
-          }
-          debug(rec);
-          if (hasETB !== -1) {
-            // send a new ACK
-            var cmd = buildAckPacket();
-            hidDevice.send(cmd, function () {
-                // we need to re-set timeout for each packet, as the number of packets are of variable length
-                abortTimer = setTimeout(function () {
-                  clearInterval(listenTimer);
-                  debug('TIMEOUT');
-                  callback('TIMEOUT', null);
-                }, timeout);
-                debug('New ACK SENT');
-            });
-          }
-          else if(rec.TYPE == ASCII_CONTROL.TER){
+        var packet = new Uint8Array(raw);
+
+        var packetHead = struct.unpack(packet, 0, '3Z2b', ['HEADER', 'SIZE', 'BYTE1']);
+
+        if(packetHead['HEADER'] != MAGIC_HEADER){
+          debug('Invalid packet from Contour device');
+          callback('INVALID_USB_PACKET');
+        }
+
+        // The tail of the packet starts 6 from the end, but because we haven't stripped the
+        // MAGIC_HEADER and length byte from packet, we're using SIZE - 2
+        var packetTail = struct.unpack(packet, parseInt(packetHead['SIZE']) - 2, '2b2Z2Z', ['CR', 'FRAME_TYPE', 'CHECKSUM', 'CRLF']);
+
+        // HID_PACKET_SIZE - 4, because we don't include the MAGIC_HEADER or the SIZE
+        if( packetHead['SIZE'] < ( HID_PACKET_SIZE - 4 ) ||
+            packetHead['BYTE1'] == ASCII_CONTROL.EOT ||
+            packetHead['BYTE1'] == ASCII_CONTROL.ENQ ||
+            packetTail['FRAME_TYPE'] == ASCII_CONTROL.ETX ||
+            packetTail['FRAME_TYPE'] == ASCII_CONTROL.ETB ) {
             clearTimeout(abortTimer);
             clearInterval(listenTimer);
-            callback(null, '');
-          }
-        });
-      }, 20); //we need to be ready for the next packet quickly
+            callback(null);
+        }
+      });
+    }, 20); // The Contour timeout is 15 seconds, but let's be slow enough to be kind to the CPU.
```

Comment:
Maybe make this comment a bit clearer that the 20 is in milliseconds, not seconds?